dsgrid/query/models.py

```diff
@@ -0,0 +1,305 @@
+
+    dataset_id: str = Field(title="dataset_id", description="Dataset ID")
+    filters: List[Any] = Field(
+        # Union[
+        #     DimensionFilterValueModel,
+        #     DimensionFilterExpressionModel,
+        #     DimensionFilterExpressionRawModel,
+        #     DimensionFilterColumnOperatorModel,
+        # ]
+        # ] = Field(
+        title="filters",
+        description="Dimension filters to apply to the dataset'",
+    )
+
+
+class MetricReductionModel(DSGBaseModel):
+    """Reduces the metric values for each set of unique dimension combinations, excluding
+    query_name."""
+
+    query_name: str = Field(
+        title="query_name", alias="dimension_query_name", description="Dimension query name"
+    )
+    operation: str = Field(
+        title="operation",
+        description="Operation to perform on the dimension.",
+    )
+
+    @validator("operation")
+    def check_operation(cls, operation):
+        allowed = ["avg", "min", "mean", "max", "sum"]
+        if operation not in allowed:
+            raise ValueError(f"operation={operation} is not supported. Allowed={allowed}")
+        return operation
+
+
+class AggregationModel(DSGBaseModel):
+
+    group_by_columns: List[str] = Field(
+        title="group_by_columns", description="Columns on which to group"
+    )
+    aggregation_function: Any = Field(
+        title="aggregation_function",
+        description="Must be a function name in pyspark.sql.functions",
+    )
+    name: Optional[str]
+
+    def dict(self, *args, **kwargs):
+        data = super().dict(*args, **kwargs)
+        data["aggregation_function"] = data["aggregation_function"].__name__
+        return data
+
+    @validator("group_by_columns")
+    def check_group_by_columns(cls, group_by_columns):
+        if not group_by_columns:
+            raise ValueError("group_by_columns cannot be empty")
+        return group_by_columns
+
+    @validator("aggregation_function")
+    def check_aggregation_function(cls, aggregation_function):
+        if isinstance(aggregation_function, str):
+            aggregation_function = getattr(F, aggregation_function, None)
+            if aggregation_function is None:
+                raise ValueError(f"{aggregation_function} is not defined in pyspark.sql.functions")
+        return aggregation_function
+
+
+class ChainedAggregationModel(DSGBaseModel):
+    aggregations: List[AggregationModel]
+
+    @validator("aggregations")
+    def check_aggregations(cls, aggregations):
+        length = len(aggregations)
+        if len(aggregations) < 2:
+            raise ValueError(f"length of ChainedAggregationModel must be at least: {length}")
+        if aggregations[-1].name is None:
+            raise ValueError("ChainedAggregationModel requires its last model to define a name.")
+        return aggregations
+
+
+# TODO: Deserializing JSON into QueryModel probably won't work for all cases, notably the unions.
+
+
+class ProjectConstraintsModel(DSGBaseModel):
```

Comment:
This name isn't super intuitive to me.

Reply:
This model defines the constraints that will be used to transform the project's data. This is separate from the parameters that define how to manipulate and store the results. This model uniquely identifies a query and so is used for caching.

I'm open to suggestions on a better name.

Reply:
ProjectQueryModel?

Reply:
`ProjectQueryModel` is the current winner